keanu-project/src/main/java/io/improbable/keanu/util/graph/AbstractGraph.java

```diff
@@ -0,0 +1,84 @@
+package io.improbable.keanu.util.graph;
+
+import java.awt.*;
+import java.util.*;
+import java.util.function.Predicate;
+
+/**
+ * This class represents a Graph (collection of nodes and edges)
+ * @param <N> The nodes of the graph
+ * @param <E> The edges of the graph
+ */
+public abstract class AbstractGraph<N extends GraphNode, E extends GraphEdge<N>> {
+
+    protected Map<String, String> metadata = new HashMap<>();
+    protected Set<E> edges = new HashSet<>();
+
+    /**
+     * @return the metadata collection for the graph
+     */
+    public Map<String, String> getMetaData() {
+        return metadata;
+    }
+
+    /**
+     * @return All the nodes in the graph
+     */
+    public abstract Collection<N> getNodes();
+
+    /**
+     * Method called before any writing is done - this can be used for lazy formatting
+     */
+    public void prepareForExport(){
+
+    }
+
+    /**
+     * @return All the edges in the graph
+     */
+    public Collection<E> getEdges() {
+        return edges;
+    }
+
+    /**
+     * This converts a java color to a .dot format
+     * @param c The color to convert
+     * @return the colot as a 6 digit hex string (e.g. #FF00FF)
+     */
+    public static final String formatColorForDot(Color c) {
```

Comment:
Does this DOT related method have to live on this very general AbstractGraph class?

Reply:
I think it's misnamed - it's actually formatting a color for HTML and DOT - and I think it makes sense to have a standard way to store colors as strings in the AbstractGraph.
I'll rename it.